source/base/parameter_handler.cc

```diff
@@ -686,7 +686,29 @@ namespace
           }
         else
           {
-            try
+            if (skip_undefined)
+              {
+                try
+                  {
+                    // it must be a subsection
+                    prm.enter_subsection(demangle(p.first), !skip_undefined);
+                    read_xml_recursively(p.second,
+                                         (current_path.empty() ?
+                                            p.first :
+                                            current_path + path_separator +
+                                              p.first),
+                                         path_separator,
+                                         patterns,
+                                         skip_undefined,
+                                         prm);
+                    prm.leave_subsection();
+                  }
+                catch (const ParameterHandler::ExcEntryUndeclared &)
+                  {
+                    // ignore undeclared entry assert
+                  }
+              }
+            else
               {
                 // it must be a subsection
                 prm.enter_subsection(demangle(p.first), !skip_undefined);
```

Comment:
This duplicates a dozen lines of code. Could you instead just put the following into the `catch` block:
```
  catch (...as before...) {
    if (skip_undefined)
    {
       // it's ok -- we are told to ignore this case, so just absorb the exception
    }
    else
    {
       // no, not ok -- we are told to report this as an error, so re-throw the exception
       throw;
    }
}
```
   
```